File: src/main/java/com/github/timo_reymann/composer_dump_autoload_phpstorm_plugin/task/ComposerAutoloadDumpRunner.java
package com.github.timo_reymann.composer_dump_autoload_phpstorm_plugin.task;

import com.github.timo_reymann.composer_dump_autoload_phpstorm_plugin.command.ComposerCommandExecutor;
import com.github.timo_reymann.composer_dump_autoload_phpstorm_plugin.util.MessageBusUtil;
import com.intellij.notification.NotificationType;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.progress.ProgressManager;
import com.intellij.openapi.project.Project;
import com.jetbrains.php.composer.ComposerDataService;
import com.jetbrains.php.composer.execution.ComposerExecution;
import com.jetbrains.php.composer.execution.executable.ExecutableComposerExecution;
import com.jetbrains.php.composer.execution.phar.PharComposerExecution;

import java.io.File;

import static com.github.timo_reymann.composer_dump_autoload_phpstorm_plugin.command.ComposerCommandExecutor.DUMP_AUTOLOAD;

/**
 * Runner for dumping autoload
 */
public class ComposerAutoloadDumpRunner {
    private final Project project;
    private final Logger logger = Logger.getInstance(ComposerAutoloadDumpRunner.class);

    public ComposerAutoloadDumpRunner(Project project) {
        this.project = project;
    }

    public void execute() {
        logger.info("Triggered autoload dump");

        // Gather needed information
        ComposerDataService composerDataService = ComposerDataService.getInstance(project);
        String workingDirectory = new File(composerDataService.getConfigPath()).getParent();
        ComposerExecution execution = composerDataService.getComposerExecution();
        ComposerCommandExecutor executor;

        // Find executor for composer to run command with
        if (execution instanceof ExecutableComposerExecution) {
            executor = new ComposerCommandExecutor(((ExecutableComposerExecution) execution).getExecutablePath(), workingDirectory, DUMP_AUTOLOAD);
        } else if (execution instanceof PharComposerExecution) {
            executor = new ComposerCommandExecutor("php", workingDirectory, ((PharComposerExecution) execution).getPharPath() + " " + DUMP_AUTOLOAD);
        } else {
            MessageBusUtil.showMessage(NotificationType.ERROR, "Unsupported composer type", "Composer runtime type is not supported!");
            return;
        }

        // Schedule task with fancy progress bar
        ProgressManager.getInstance()
                .run(new ComposerCommandBackgroundTask(project, "Dump autoload", executor));
    }
}

File: src/main/kotlin/de/timo_reymann/composer_dump_autoload_phpstorm_plugin/task/ComposerAutoloadDumpRunner.kt
package de.timo_reymann.composer_dump_autoload_phpstorm_plugin.task

import com.intellij.notification.NotificationType
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.project.Project
import com.jetbrains.php.composer.ComposerDataService
import com.jetbrains.php.composer.execution.executable.ExecutableComposerExecution
import com.jetbrains.php.composer.execution.phar.PharComposerExecution
import de.timo_reymann.composer_dump_autoload_phpstorm_plugin.command.ComposerCommandExecutor
import de.timo_reymann.composer_dump_autoload_phpstorm_plugin.settings.ComposerDumpAutoloadSettings
import de.timo_reymann.composer_dump_autoload_phpstorm_plugin.util.MessageBusUtil
import java.io.File

/**
 * Runner for dumping autoload
 */
class ComposerAutoloadDumpRunner(private val project: Project) {
    private val logger = Logger.getInstance(
        ComposerAutoloadDumpRunner::class.java
    )

    private fun getArgs(): String {
        val parameters = ComposerDumpAutoloadSettings.getInstance(project).arguments
        val delimiter = when {
            parameters.isBlank() -> ""
            else -> " "
        }
        return "${ComposerCommandExecutor.DUMP_AUTOLOAD}${delimiter}${parameters}"
    }

    fun execute() {
        logger.info("Triggered autoload dump")

        // Gather needed information
        val composerDataService = ComposerDataService.getInstance(project)
        val workingDirectory = File(composerDataService.configPath).parent
        val args = getArgs()

        // Find executor for composer to run command with
        val executor = when (val execution = composerDataService.composerExecution) {
            is ExecutableComposerExecution -> ComposerCommandExecutor(
                execution.executablePath,
                workingDirectory,
                args
            )

            is PharComposerExecution -> ComposerCommandExecutor(
                "php",
                workingDirectory,
                "${execution.pharPath} $args"
            )

            else -> {
                MessageBusUtil.showMessage(
                    project,
                    NotificationType.ERROR,
                    "Unsupported composer type",
                    "Composer runtime type is not supported!"
                )
                return
            }
        }

        // Schedule task with fancy progress bar
        ProgressManager.getInstance()
            .run(ComposerCommandBackgroundTask(project, "Dump autoload", executor))
    }
}
